Feat: Fetcher Plugin System (#277)
SE-UUlm/snowballr-backend: wiki/Fetcher.md

> [!WARNING]
> Python fetchers execute on the same system as the backend instance and thus
> have direct access to the same system. They can access the filesystem, create
> network connections and much more. ONLY USE TRUSTED FETCHERS!

There are a lot of online databases providing access to a vast collection of
scientific publications. SnowballR aims to unite these online databases and
integrate them into a coherent and low-friction user experience. Instead of
time-consumingly needing to search these databases by hand, SnowballR takes
care of it behind the scenes so you can focus on the important part: reviewing

> papers.

Reviewer: ```suggestion
care of it behind the scenes and provide the user with a collection of relevant papers that they can review directly.
```

> To access these services, the SnowballR backend needs a way to talk to them.
> This could for example be an http client sending requests to a REST API. An

Reviewer: ```suggestion
To access these databases, the SnowballR backend needs a way to talk to them.
This could, for example, be an http client sending requests to a REST API of the paper database. An
```

Reviewer: And please always replace services by databases as services could be a bit confusing imo. Or do we request from something different, I do not think so?!

Author: I think it is rather difficult to find a suiting term.
What is confusing about _services_?

I could see how _database_ could already mean two different things in this context: _database_ as in "PostgreSQL and SQLite" or _database_ as in "a website serving a collection of data (for example papers)". The latter often employs the former, yet it is more of a web service in my opinion.

Technically, this system allows accessing papers in an arbitrary way, not just classical databases. I think _services_ is a broader term and suits the desired meaning better, but I am open to discussion.

Reviewer: No, i understand your point with the different interpretations of _database_ as well. I thought it could be confusing, because when talking about services, we often speak of our project service and user service etc. and the fetcher requests more like data or knowledge bases, but lets keep it that way.

object allowing SnowballR to interface with a database will be referred to as
a _fetcher_.

As the landscape of services is ever-changing, hard-coding a fetcher for each

> and every one of these inside of SnowballR's backend would be very costly and

Reviewer: ```suggestion
As the landscape of services is ever-changing, hard-coding a fetcher for every
one of these inside the SnowballR backend would be very costly and
```

fragile. Yet, SnowballR should not be your limiting factor. It should adapt and
evolve; be customizable to suit your needs. That is why we decided to implement

> a plugin system, allowing you to create your very own fetchers with ease.

Reviewer: ```suggestion
fragile. To allow adaptions and evolution, we decided to implement
a plugin system, which allows to create own fetchers with ease.
```

Author: Why the change? This is a preface for a page intended to be read mostly by users, not project developers. I think a bit more playful language is ok here.

Reviewer: But users are not directly end users, but developers or persons with technical background, otherwise they couldn't write their own fetchers and setup everything. But if you want, keep it.

At the heart of this plugin system is the [Python](https://www.python.org/)
programming language. It has a low barrier of entry and easy syntax whilst also
being powerful and developer-friendly. Furthermore, it provides a broad set of
builtin functionality and is very widespread, even in non-computer-science
degrees. This makes it a very natural and favourable choice.

> [!IMPORTANT]
> Installing fetchers requires direct access to a running instance of the
> SnowballR backend.

The following guide assumes you have basic python knowledge and direct access
to a running SnowballR backend instance.

### Prerequisites

> - A working Python installation.

Reviewer: Please specify the required python version

Author: Great suggestion 👍 I've added the version requirement.

- A properly set-up and activated [Python virtual environment](https://docs.python.org/3/library/venv.html).

> - Your favourite code editor.

Reviewer: The entire thing around the code editor and similar stuff (VSCode, ruff, ...) is not necessary. Everyone ones that you need some kind of editor to edit files and I know nobody that uses Zed or ruff. And as you stated: basic python knowledge is required, which includes imo how to setup the language server.

Author: This is a page intended mainly for users. It's think it is fine. Furthermore, a little tip for increasing the developer experience cannot hurt.

Yes, someone who can develop should know that you need a text editor (although not true in every case), but even then, having this hint is kind of cool. Not everyone is highly experienced or has their setup fully worked out 🤷.

Just because _you_ don't know anybody that actively uses a technology doesn't invalidate it. Both are huge projects with a big community.

Reviewer: Yeah, but why did you leave out PyCharm or NeoVim, both very large projects with probably even larger communities as well? Maybe we have too different opinions on that, but imo a wiki should help you set up what is necessary for this project and not what code editors are nice. 

But I leave it to you to make the change or close this conversation on your own.

The plugin system requires a base set of python dependencies to work. Install them
using the following command (again, make sure you're in the same venv the snowballR
backend will later be executed in):

```

> pip install requests dataclass_wizard pycurl

Reviewer: Can't you provide a requirements.txt as in the pipeline? I think this would be easier, especially if we extend this system and the list of required packages grows. And you could directly use this file in the Pipeline.

Author: The `requirements.txt` is a great idea. I've added it to the project. However, this guide cannot make full use of it. I added a link to the up-to-date `requirements.txt` file, but I'd keep the package names for now, as the guide does not require having a source copy of the SnowballR backend.

Reviewer: Probably more complicate to keep the documentation up to date this way, but it is also ok.

directory is relative to the working directory of the backend and should
already be present if you have previously started the backend. Every `*.py`
file directly contained within the `fetchers` directory will be treated as a
fetcher. If you would like to create a reusable module/library, put it inside a
subdirectory like the `lib` folder or create a new one.

### Fetcher Contract

Every fetcher is required to follow an implicit contract to be compatible with
the backend. This contract is automatically adhered to if you use the bundled
SnowballR library:

```py
from SnowballR import FetcherPlugin, Paper

availableOptions: set[str] = ...
def searchPapers(searchQuery: str, options: dict[str, str]) -> set[Paper]: ...
def fetchForwardReferences(paper: Paper, options: dict[str, str]) -> set[Paper]: ...
def fetchBackwardReferences(paper: Paper, options: dict[str, str]) -> set[Paper]: ...

FetcherPlugin(
    availableOptions,
    searchPapers,
    fetchForwardReferences,
    fetchBackwardReferences,
).start()
```

By using `FetcherPlugin` and calling `.start()` on it, everything will be taken
care of. You just have to provide implementations for the functions and you're

> off to go.

Reviewer: ```suggestion
care of. Just provide implementations for the functions, and you're off to go.
```

Author: Why is your suggestion better?

Reviewer: No "You", but as you already noticed, I am not a fan of such a "I-am-your-friend-and-we-talk-a-bit" documentation style but prefer more technical documentations. Additionally, it is a _suggestion_ and can be ignored (close on your own).

To accommodate for API secrets and other variables, each fetcher is configurable
using a string-to-string dictionary. The `availableOptions` set serves as a hint
to the frontend, which options this fetcher accepts. There could, however, at
any time be missing or additional entries. Project admins can select a fetcher
and provide key-value pairs as options in the project settings. Remember during

> the implementation: A fetcher can be used accross multiple projects, each with

Reviewer: ```suggestion
the implementation: A fetcher can be used across multiple projects, each with
```

Reviewer: But I do know quite understand what you want me to tell with this section.

Author: This section introduces `availableOptions` and how they're used within the project.

their own set of options.

### Writing a Fetcher

Equipped with this knowledge, let's get started and write our
own example fetcher. SnowballR provides a predefined[^1] `Paper`
dataclass, which is expected as a result. The source is located in
`./plugins/fetchers/lib/SnowballR/fetcher_plugin.py` and can be imported like

> this: `from SnowballR import Paper`. It is mutable and can be constructed like

Reviewer: ```suggestion
`./plugins/fetchers/lib/SnowballR/fetcher_plugin.py` and can be imported with
`from SnowballR import Paper`. It is mutable and can be constructed like
```

this:

```py
paper = Paper(
    title,            # str (required)

>     abstract,         # str (required)

Reviewer: An abstract should not be required as well as the set of authors.

Author: Every field except for `externalId` is required. This should align with the database table definition for papers. I've corrected the doc regarding the replacement of `publishedAt` with `year`, though.

Reviewer: Ok, but then a short hint would be helpful, explaining what to do in case your fetched paper has no abstract (I guess we want to default to the empty string?)

    fetcherMetadata,  # map[str, str] (required)
    authors,          # set[Author] (required)
    externalId,       # str
    publishedAt,      # int: seconds since epoch
    publisher,        # str
    publicationType,  # str
    publicationName   # str
)
```

> Similarily, the `Author` can be constructed like this:

Reviewer: ```suggestion
Similarly, the `Author` can be constructed like this:
```

```py
author = Author(
    firstName, # str (required)
    lastName,  # str (required)
    orcid      # str
)
```

> Let's put all the pieces together:

Reviewer: I think such sentences are not necessary, but it would be better to explain in the next section what your basic fetcher does: It is a fetcher without any options, that always finds a test paper with the title "title" no matter what the search string is, and that can neither fetch forward nor backward references.

Author: I've kept the playful language but also added an explanation of what the fetchers do, good suggestion 👍.

Reviewer: Thank you.

#### Basic Fetcher

```py
# ./plugins/fetchers/basic.py

from SnowballR import FetcherPlugin, Paper

availableOptions = {}

def searchPapers(searchQuery: str, options: dict[str, str]) -> set[Paper]:
    return { Paper("title", "abstract") }

def fetchForwardReferences(paper: Paper, options: dict[str, str]) -> set[Paper]:
    return {}

def fetchBackwardReferences(paper: Paper, options: dict[str, str]) -> set[Paper]:
    return {}

FetcherPlugin(
    availableOptions,
    searchPapers,
    fetchForwardReferences,
    fetchBackwardReferences,
).start()
```

Of course, this is not very useful. A common way of acquiring data is using some

> sort of REST API. Let's take a look at how one would do that.

Reviewer: ```suggestion
sort of REST API. 
```

Author: I am in favor of these kinds of sentences.

#### HTTP Requests

We'll make use of the previously installed `requests` package to acquire
external resources.

```py

> # ./plugins/fetchers/http.py

Reviewer: What is the purpose of this comment?

Author: It indicates where this file lives.

Reviewer: Ah ok.

from SnowballR import FetcherPlugin, Paper
import requests

availableOptions = {
    "API_KEY"
}

def searchPapers(searchQuery: str, options: dict[str, str]) -> set[Paper]:
    content = requests.get(
        "https://example.com",
        auth=("user", options["API_KEY"])
    ).text

    return {
        Paper("title", content)
    }

def fetchForwardReferences(paper: Paper, options: dict[str, str]) -> set[Paper]:
    return {}

def fetchBackwardReferences(paper: Paper, options: dict[str, str]) -> set[Paper]:
    return {}

FetcherPlugin(
    availableOptions,
    searchPapers,
    fetchForwardReferences,
    fetchBackwardReferences,
).start()
```

But the possibilities do not end here. You could access a filesystem database or
resort to many pip/python libraries.

#### Debugging

You can use the global `log` instance to print messages to the backend's log.
The interface looks like this:

```py
class Logger:
    def info(self, msg: str): ...
    def warn(self, msg: str): ...
    def error(self, msg: str): ...
    def debug(self, msg: str): ...
    def trace(self, msg: str): ...
```

> Just use it anywhere in your code, ensuring that it is already imported:

Reviewer: ```suggestion
Just use it anywhere in your code, but ensure that it is already imported:
```

Reviewer: I think you forgot this (and this is not a critic to your playful language, but I think your sentence means something different than intended).

object allowing SnowballR to interface with a database will be referred to as
a _fetcher_.

As the landscape of services is ever-changing, hard-coding a fetcher for every
one of these inside the SnowballR backend would be very costly and fragile. Yet,
SnowballR should not be your limiting factor. It should adapt and evolve; be
customizable to suit your needs. That is why we decided to implement a plugin
system, allowing you to create your very own fetchers with ease.

At the heart of this plugin system is the [Python](https://www.python.org/)
programming language. It has a low barrier of entry and easy syntax whilst also
being powerful and developer-friendly. Furthermore, it provides a broad set of
builtin functionality and is very widespread, even in non-computer-science
degrees. This makes it a very natural and favourable choice.

> [!INFO]
> Installing fetchers requires direct access to a running instance of the
> SnowballR backend. Contact your instance administrator if required.

The following guide assumes you have basic python knowledge and direct access
to a running SnowballR backend instance.

### Prerequisites

> - A working Python installation (Version >= 3.10).

Reviewer: ```suggestion
- A working Python installation (Version >= 3.12).
```

In the CI you use 3.13 and in the `snowballr.py` some features from 3.12.

Author: The last time I checked it was `3.10`, but the files have changed since then. Thank you 👍

- A properly set-up and activated [Python virtual environment](https://docs.python.org/3/library/venv.html).
- Your favourite code editor.

The plugin system requires a base set of python dependencies to work. Install
them using the following command (again, make sure you're in the same `venv` the
SnowballR backend will later be executed in):

```bash
pip install -r requirements.txt
```

The list of required packages can be also found in the [requirements.txt](https://github.com/SE-UUlm/snowballr-backend/blob/develop/requirements.txt).

> [!IMPORTANT]
> The plugin system makes use of the environments `python` installation. Again,
> make sure to execute the backend in the correct `venv`, where the required
> packages were also installed.

To get proper autocomplete and type checking for the SnowballR types, add the

> `./plugins/fetchers/lib/` directory to your python path. This can be done by

Reviewer: Here you are inconsistent to your implementation. As I already suggested in another conversation, I think it would be great to move the plugin files into a fetcher directory, however, if you keep the current version, adapt the documentation (I think the most important parts are already corrected in the commit [c71c38715f7c3006911cb6236469ed09326d3a76](https://github.com/SE-UUlm/snowballr-backend/commit/c71c38715f7c3006911cb6236469ed09326d3a76)

Author: As I have changed the root, this should be correct again.

the backend. This contract is automatically adhered to if you use the bundled
SnowballR library:

```py

from snowballr import fetcher_plugin, Paper

options: dict[str, str] = ...
def search_papers(search_query: str, options: dict[str, str]) -> list[Paper]: ...
def forward_references(paper: Paper, options: dict[str, str]) -> list[Paper]: ...
def backward_references(paper: Paper, options: dict[str, str]) -> list[Paper]: ...

fetcher_plugin(
    options,
    search_papers,
    forward_references,
    backward_references,
)
```

Everything regarding the protocol will be taken care of by calling `fetcher_plugin`.
Just provide implementations for the functions and you're off to go.

To accommodate for API secrets and other variables, each fetcher is configurable
using a string-to-string dictionary. The `options` set serves as a hint
to the frontend, which options this fetcher accepts. There could, however, at
any time be missing or additional entries. Project admins can select a fetcher
and provide key-value pairs as options in the project settings. Remember during
the implementation: A fetcher can be used across multiple projects, each with
their own set of options.

> ### Writing a Fetcher

Reviewer: One thing: If you want to permanently add a fetcher to the project (so for developers, not users of snowballr), then you also need to "register" this fetcher in the PythonFetcherManager.kt or so, right? Maybe this could be mentioned as well?

Author: I added a section `Developers` which briefly states this. 👍

Of course, this is not very useful. A common way of acquiring data is using some
sort of REST API. Let's take a look at how one would do that.

#### HTTP Requests

We'll make use of the installed `requests` package (which is already a
dependency) to acquire external resources.

```py
# ./plugins/fetchers/http.py

from snowballr import fetcher_plugin, Paper, Author
import requests

options = {
    "API_KEY": "The API key"
}

def search_papers(searchQuery: str, options: dict[str, str]) -> list[Paper]:
    content = requests.get(
        "https://example.com",
        auth=("user", options["API_KEY"])
    ).text

    return [ Paper(
        "title",
        None,
        content,
        2026,
        "publisher",
        "publication_type",
        "publication_name",
        [Author("first_name", "last_name")],
        {}
    ) ]

def forward_references(paper: Paper, options: dict[str, str]) -> list[Paper]:
    return []

def backward_references(paper: Paper, options: dict[str, str]) -> list[Paper]:
    return []

fetcher_plugin(
    options,
    search_papers,
    forward_references,
    backward_references,
)
```

> This fetcher accepts an `API_KEY` option, displaying `hint` in the frontend

Reviewer: is `hint` correct here? Where does it come from?

Author: Accidental mistake. Corrected it. 👍
